---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# MutTranscript
Analyzing Transcript-level Properties of Mutation Data in VCF Files

<!-- badges: start -->
<!-- badges: end -->

## Description

The tool will allow users to analyze mutation data from
variant call format (VCF) files. The purpose of this tool is to streamline 
the analysis of transcription-related effects of mutations, such as analyzing 
whether the mutations are synonymous or nonsynonymous at the amino acid level, 
whether they have any deleterious effects and so on. The tool also enables 
the visualization of mutations in exonic regions, allowing the exploration 
of how these transcription-related effects vary along the length of a gene. 
It has the functionality to map the genomic location of mutations to 
their location on the coding sequence (CDS) of the transcript, with the introns 
removed.

Although there are already many existing tools to annotate mutation data from
sequencing, some of which this package relies on, no package exists for the
specific purpose of visualizing mutation property distribution and how it varies
with the distance along genes. This information can be relevant for, for example,
uncovering biological effects such as mutation bias on particular ends of genes or
technical effects such as bias towards the 3' or 5' ends of genes in particular
RNA sequencing platforms. Therefore, we hope that this package will enable exploratory
analysis of such effects and thus guide further downstream analysis of mutation data.

This package was built using R version 4.4.2 on an aarch64-apple-darwin20 platform, 
running under macOS Ventura 13.2.1.

## Installation

You can install the development version of MutTranscript like so:

``` r
install.packages("devtools")
library("devtools")
devtools::install_github("jahink17/MutTranscript", build_vignettes = TRUE)
library("MutTranscript")
```

## Overview

## Contributions

## References

```r
ls("package:<PackageName>")
data(package = "<PackageName>") # optional
browseVignettes("<PackageName>")
```

## Acknowledgements

This package was developed as part of an assessment for 2024 BCB410H: Applied Bioinformatics
course at the University of Toronto, Toronto, CANADA. `MutTranscript` welcomes issues,
enhancement requests, and other contributions. To submit an issue, use the GitHub issues.

## Example

This is a basic example which shows you how to solve a common problem:

```{r example}
library(MutTranscript)
## basic example code
```

What is special about using `README.Rmd` instead of just `README.md`? You can include R chunks like so:

```{r cars}
summary(cars)
```

You'll still need to render `README.Rmd` regularly, to keep `README.md` up-to-date. `devtools::build_readme()` is handy for this.

You can also embed plots, for example:

```{r pressure, echo = FALSE}
plot(pressure)
```

In that case, don't forget to commit and push the resulting figure files, so they display on GitHub and CRAN.
